cli/config/credentials/default_store.go

```diff
@@ -2,21 +2,70 @@ package credentials
 
 import "os/exec"
 
-// DetectDefaultStore return the default credentials store for the platform if
-// no user-defined store is passed, and the store executable is available.
-func DetectDefaultStore(store string) string {
-	if store != "" {
+// DetectDefaultStore returns the credentials store to use. If no user-defined
+// custom helper is passed.
+//
+// Some platforms define a preferred helper, in which case it attempts to look
+// up the helper binary before falling back to the platform's default.
+//
+// If no user-defined helper is passed, and no helper is found, it returns an
+// empty string, which means credentials are stored unencrypted in the CLI's
+// config-file without the use of a credentials store.
+func DetectDefaultStore(customStore string) string {
+	if customStore != "" {
 		// use user-defined
-		return store
+		return customStore
 	}
```

Comment:
Right, so I the logic here was that if the config-file has a credentials-store configured, we accept it whatever it is, and consider it an option set by the user; when used, it may error, but that's intentional (you configured a helper, we tried using it, and it failed! intervention is needed).

Reply:
But now the fun bit; https://github.com/docker/cli/blob/32ff200fe6d843843718709ffa94023eb7356108/cli/config/config.go#L162-L172

See what's wrong in the logic there? Hint: https://github.com/docker/cli/blob/32ff200fe6d843843718709ffa94023eb7356108/cli/config/config.go#L168


That nice porcelain `ContainsAuth()` function checks if either a credentials-store is configured or per-registry helpers, or contains credentials; https://github.com/docker/cli/blob/32ff200fe6d843843718709ffa94023eb7356108/cli/config/configfile/file.go#L90-L94

I guess we could consider it an extra line of defence to never overwrite what's there, but definitely a bit iffy, and looks like it was there from the start https://github.com/moby/moby/commit/cf721c23e715e545eccf8484e145c2d18d6a6a23